### PR TITLE
torchgen supports generate the code that for "Register{Privateuse1}.cpp"

### PR DESCRIPTION
**Pull request by huihoaan**
In my last pr https://github.com/pytorch/pytorch/pull/99406, I can rename the DispatchKey.PrivateUse1 to my custom device(suppose its name is YPU).
But in the function `gen_dispatcher_registrations`:
https://github.com/pytorch/pytorch/blob/675029aabf143a60aee62c21e093ed1091f46743/torchgen/gen_backend_stubs.py#L456-L466
Line 464 will pass YPU to the CodeTemplate, and will generator the following code:
```
TORCH_LIBRARY_IMPL(aten, YPU, m) {
    $dispatch_registrations_body
};
```
Obviously, this code cannot register the operators to pytorch.
Now, the problem is that for DispatchKey.PrivateUse1, it needs to have two representations of str. Most of the time, it should be "YPU", but when the operator is actually registered, it needs to be made PrivateUse1.
So, I want add a new parameter to `gen_dispatcher_registrations` for real device.

### CODE REVIEW

**Comment by huihoaan** [comment]
@ezyang  Can you take a look at this pr, give some advices or other solution?

**Comment by ezyang** [comment]
TBH, my preference would be to revert the previous patch we landed for this and figure out some other strategy. It's just native_functions.yaml parsing you want to modify right? Maybe we can keep PrivateUse1 non-monkeypatched and only do the translation on input reading.

**Comment by huihoaan** [comment]
> TBH, my preference would be to revert the previous patch we landed for this and figure out some other strategy. It's just native_functions.yaml parsing you want to modify right? Maybe we can keep PrivateUse1 non-monkeypatched and only do the translation on input reading.

OK， let me think about how do I do both at the same time.

File: torchgen/model.py [hunk 1]
```diff
@@ -187,13 +187,19 @@ class DispatchKey(Enum):
     # END autogenerated
 
     def __str__(self) -> str:
+        return self.name.replace("PrivateUse1", PrivateUse1Name.custom_name)
+
+    # for PrivateUse1, it needs to have two representations of str.Most of the time, it should be the name which user
+    # set, but when the operator is actually registered, it needs to be made PrivateUse1.
```
Comment (ezyang) [inline torchgen/model.py:193]: Are you sure? It seems to me like most of the time you want PrivateUse1 when this dispatch key is printed throughout torchgen

**Comment by ezyang** [comment]
This seems better but I would prefer if we don't change the str behavior. Can you explain why you want that?

**Comment by huihoaan** [comment]
> This seems better but I would prefer if we don't change the str behavior. Can you explain why you want that?

In our project, we will provide the user with `{DispatcherKey}NativeFunctions.h`. If we don't change the str behavior, it will be `PrivateUse1NativeFunctions.h` and the code in this file will also be about `PrivateUse1`. This will confuse our users and make no sense.

File: torchgen/model.py (hunk 1 above)
Comment (huihoaan) [reply torchgen/model.py:193]: I'm sure, because of our custom device has its own name(e.g. YPU), it's better to be 'YPU' in this code generated with torchgen. But, when I register our operator to my custom device, its str behavior must be "PrivateUse1" rather than "YPU".

**Comment by ezyang** [comment]
If it is only this site you want changed, can you surgically only do the translation there?

**Comment by huihoaan** [comment]
@ezyang I removed the method added in the last commit and just translated it before generating the registration operator code.

File: torchgen/model.py [hunk 2]
```diff
@@ -187,13 +187,14 @@ class DispatchKey(Enum):
     # END autogenerated
 
     def __str__(self) -> str:
-        return self.name
+        return self.name.replace("PrivateUse1", PrivateUse1Name.custom_name)
```
Comment (ezyang) [inline torchgen/model.py:190]: No. I don't want this patched. Revert this change, and then modify the call sites you want translated one by one.
